File: cuda-numba/axpy.py
from math import sin, cos

import numpy
import numba
import numba.cuda

from _axpy import daxpy

n = 10000
a = 3.4

# initialise data and calculate reference values on CPU
x = numpy.array([sin(i) * 2.3 for i in range(n)], float)
y = numpy.array([cos(i) * 1.1 for i in range(n)], float)
y_ref = a * x + y

# allocate + copy initial values
x_ = numba.cuda.to_device(x)
y_ = numba.cuda.to_device(y)

# calculate axpy on GPU
daxpy(n, a, x_.device_ctypes_pointer.value, y_.device_ctypes_pointer.value)

# copy result back to host and print with reference
print('  initial: {0} {1} {2} {3} {4} {5}'.format(
    y[0], y[1], y[2], y[3], y[-2], y[-1]))
y_.copy_to_host(y)
print('reference: {0} {1} {2} {3} {4} {5}'.format(
    y_ref[0], y_ref[1], y_ref[2], y_ref[3], y_ref[-2], y_ref[-1]))
print('   result: {0} {1} {2} {3} {4} {5}'.format(
    y[0], y[1], y[2], y[3], y[-2], y[-1]))

File: hipblas/axpy.c
#include <stdio.h>
#include <math.h>
#include <hip/hip_runtime.h>
#include <hipblas.h>

int main(void)
{
    int i;
    const int n = 100;
    double a = 3.4;
    double x[n], y[n], y_ref[n];
    double *x_, *y_;
    hipblasHandle_t handle;

    // initialise data and calculate reference values on CPU
    for (i=0; i < n; i++) {
        x[i] = sin(i) * 2.3;
        y[i] = cos(i) * 1.1;
        y_ref[i] = a * x[i] + y[i];
    }
    printf("  initial: %f %f %f %f ... %f %f\n",
            y[0], y[1], y[2], y[3], y[n-2], y[n-1]);
    hipblasCreate(&handle);

    // allocate + copy initial values
    hipMalloc((void **) &x_, sizeof(double) * n);
    hipMalloc((void **) &y_, sizeof(double) * n);
    hipMemcpy(x_, x, sizeof(double) * n, hipMemcpyHostToDevice);
    hipMemcpy(y_, y, sizeof(double) * n, hipMemcpyHostToDevice);

    // calculate axpy on GPU
    hipblasDaxpy(handle, n, &a, x_, 1, y_, 1);

    // copy result back to host and print with reference
    hipMemcpy(&y, y_, sizeof(double) * n, hipMemcpyDeviceToHost);
    printf("reference: %f %f %f %f ... %f %f\n",
            y_ref[0], y_ref[1], y_ref[2], y_ref[3], y_ref[n-2], y_ref[n-1]);
    printf("   result: %f %f %f %f ... %f %f\n",
            y[0], y[1], y[2], y[3], y[n-2], y[n-1]);

    hipblasDestroy(handle);
    return 0;
}
